### PR TITLE
feat(node): v11.6

### LINKED ISSUE #32744: Definition for "RemoteInfo" and related functions in dgram for Node
- [X] I tried using the `@types/node` package and had problems.
- [X] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [X] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [X] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
- Authors: @rbuckton 

I didn't want to make a pull request since I'm very new to typescript and I might be missing something, but it seems there are a few issues with the type definitions for the Node dgram module (as described in https://nodejs.org/api/dgram.html)

Namely, the "RemoteInfo" interface is missing a size : number (this caused a compiler error on JS code I was already using and trying to migrate to TS which led me to finding the issue) and in several places after that (basically in all the listener configuration functions) there are references to "AddressInfo" which should be "RemoteInfo" instead.

### LINKED ISSUE #32926: @types/node Buffer
If you know how to fix the issue, make a pull request instead.
I'm not sure how, because that module is in the process of being reorganized right this instant.

- [x] I tried using the `@types/node` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @Microsoft, @DefinitelyTyped, @jkomyno, @a-tarasyuk, @alvis, @r3nya, @brunoscheufler, @smac89, @tellnes, @DeividasBakanas, @eyqs, @Flarna, @Hannes-Magnusson-CK, @KSXGitHub, @hoo29, @kjin, @ajafff, @islishude, @mwiktorczyk, @matthieusieben, @mohsen1, @n-e, @octo-sniffle, @parambirs, @eps1lon, @SimonSchick, @ThomasdenH, @WilcoBakker, @wwwy3y3, @ZaneHannanAU, @jeremiergz, @samuela, @kuehlein.

The type for `Buffer.prototype.write` is wrong, and causes mistakes.
It currently looks like this:
```ts
write(string: string, offset?: number, length?: number, encoding?: string): number;
```
Besides the fact that it uses `string` instead of the 
```ts
type BufferEncoding = "ascii" | "utf8" | "utf16le" | "ucs2" | "base64" | "latin1" | "binary" | "hex";
```
defined above it, makes it look like 
```ts
buf.write("Ы", 0, undefined, "latin1") // 2
```
will work and return 1.
Instead, having `undefined` as 3rd argument will make it ignore the encoding, which results in hard to track error.
The real type needs to be defined with overloads, because the **only** way to make it use encoding without setting an actual length is
```ts
buf.write("Ы", 0, "latin1") // 1
```
In fact, as soon as it sees a string, it thinks it's `encoding`:
```ts
buf.write("Ы", "latin1") // 1
```
The [node documentation](https://nodejs.org/api/buffer.html#buffer_buf_write_string_offset_length_encoding) doesn't specify this behavior either, so I'd assume as much as whoever wrote the types.

I only found out by trying 
```
a.write("Ы", 0, Infinity, 'latin1')
```
which returned `TypeError: Unknown encoding: Infinity`

### PR DESCRIPTION
**Pull request by SimonSchick**
Please fill in this template.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V11.md
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

This adds misc things from node v11.6, some of the changes are being worked on by others so I didn't add those, see:

In addition I've started moving module tests into the /tests folder to avoid constantly having to touch this overly large file.

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32787
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32835
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32597

@rbuckton as a note, would you be fine with removing typings for all node version that are EOL? (v0, v4, v5, v7, v9) it should prevent people from editing old/obsolete typings.

Also:
fixes #32744
fixes #32926
Partially addresses #31505

### CODE REVIEW

**Comment by SimonSchick** [comment]
@Flarna knowing the @types/node CI this will take a while to pass, feel free to leave comments :)

**Comment by SimonSchick** [comment]
Ok, having some problems with `TextEncoder`, mostly packages loading dom and node typings.

**Comment by Flarna** [comment]
I think it's a good idea to remove obsolete type definitions but I would do this in a separate PR just in case someone complains and a revert is needed. CI for node was green yesterday so maybe a good time window today.

**Comment by SimonSchick** [comment]
@Flarna problem is that `TextEncoder` conflicts with dom typings now, I don't see how I can fix this :/

**Comment by Flarna** [comment]
I think #25342 was about a similar problem with `URL` which was added to `global`. Solution that time was to simply skip this and it seems no one complained till now :o).

In general I recommend to keep using module `dom` in tests here as any incompatibilities with `dom` result in a lot broken builds in the wild.

**Comment by SimonSchick** [comment]
@Flarna yea that's what I thought too, problem is that this really isn't a sustainable solution.

**Comment by SimonSchick** [comment]
Ok, reverted `TextEncoder` changed to make sure this can be merged.

In addition I went over the deprecation log, added deprecation tags and removed dead APIs as appropriate.

**Comment by SimonSchick** [comment]
@armanio123 unrelated build failure due to:
```
Error in s3-download-stream
Error: /home/travis/build/DefinitelyTyped/DefinitelyTyped/types/s3-download-stream: Contains the word 'download', which is banned by npm.
```
I don't think this is my fault, should I remove the package or something?

**Comment by armanio123** [comment]
@SimonSchick it seems there's a new rule added by npm that's breaking the types-publisher. We are disallowing it for the moment while we have a better solution. @sandersn have a more context on this.

**Comment by SimonSchick** [comment]
@sandersn can we skip/warn on those packages instead of failing the build if it's not the primary package being tested? 😞

**Comment by SimonSchick** [comment]
Also can you provide the source on this rule change, can't find anything on this.

**Comment by sandersn** [comment]
Here's what I got from NPM support this morning:

> The word `download` is not allowed in package names so you will have to find an alternative for your package. Unfortunately, we are unable to override this.

Then, after asking for the banned word list so I could avoid trying to publish packages with those words:
> Unfortunately, we don't currently have a public list of banned words. You will see "That word is not allowed" when a word doesn't pass the spam algorithm.

This is the only npm-official reference I found to spam filtering: https://blog.npmjs.org/post/169582189317/incident-report-npm-inc-operations-incident-of. I assume they don't talk much about it because they're going for security by obscurity, or are not completely in control of their filtering software and don't know ahead of time what will be rejected.


There are only 3 packages that currently use the word 'download': 's3-download-stream', 'download' and 'downloadjs'. I assume that those packages are allowed just because they were on NPM before the rule change. I'll manually exclude them from the rule and update dtslint, because I *really* don't like random packages blocking `@types` publishing.

**Comment by SimonSchick** [comment]
That's very un-transparent of them, are there any plans to create a secondary package registry (run by DT/MS) and map an org (for examples @types2) to that registry?

Anyways, is there a tracking issue for this, I'd like to voice my opinion there!

**Comment by sandersn** [comment]
I pushed a fix to dtslint, so I'll restart the Travis build.

**Comment by SimonSchick** [comment]
👍

**Comment by SimonSchick** [comment]
Also went ahead and:
fixed #32926 in this major version as it could be considered a breaking change.
Added `SharedArrayBuffer` as class.
Enabled strict null checks.

**Comment by Flarna** [comment]
#32661 and #32661 have been merge recently and therefore these changes are missing in our v10 variant.

**Comment by SimonSchick** [comment]
@Flarna you posted the same PR twice.
Also I think I can just rebase?

**Comment by Flarna** [comment]
Ups.. corrected comment above. I don't think rebase is enough as this would not update the files in v10 folder because they don't exist on master. Git is usually not able to track something like this.

**Comment by SimonSchick** [comment]
done

File: types/node/child_process.d.ts
Comment (Flarna) [inline types/node/child_process.d.ts]: `env` is already included via `CommonOptions`

File: types/node/child_process.d.ts
Comment (Flarna) [inline types/node/child_process.d.ts]: Is `extends CommonOptions` missing here? if yes, `env`, `cwd` are already included there.

Comment (SimonSchick) [reply types/node/child_process.d.ts]: No because of `windowsHide`, I tried to dedupe a little bit here and wanted to avoid creating too many `common` interfaces.

File: types/node/crypto.d.ts
Comment (Flarna) [inline types/node/crypto.d.ts]: according to docs `DiffieHellman.setPrivateKey` and `DiffieHellman.setPublicKey` are not deprecated but `ecdh.setPublicKey` is.

Comment (SimonSchick) [reply types/node/crypto.d.ts]: Oh, we don't have that, so I'm just going to ignore it.

File: types/node/child_process.d.ts
Comment (Flarna) [reply types/node/child_process.d.ts]: Sorry, don't get this. [docs](https://nodejs.org/dist/latest-v11.x/docs/api/child_process.html#child_process_child_process_execfilesync_file_args_options) tell that `uid`, `env`, and `windowsHide` are valid options for `ExecFileSync`

File: types/node/fs.d.ts
Comment (Flarna) [inline types/node/fs.d.ts]: according to [docs](https://nodejs.org/dist/latest-v11.x/docs/api/fs.html#fs_fs_lchown_path_uid_gid_callback) `lchown` and `lchownSync` are no longer deprecated since 10.6.0

File: types/node/tty.d.ts
Comment (Flarna) [inline types/node/tty.d.ts]:
```suggestion
        getColorDepth(env?: {}): number;
```

File: types/node/index.d.ts
Comment (Flarna) [inline types/node/index.d.ts]:
```suggestion
// Type definitions for Node.js 11.8
```
`timeout` option in `tls.connect()` was added in 11.8

File: types/node/ts3.1/tsconfig.json
Comment (Flarna) [inline types/node/ts3.1/tsconfig.json]: shouldn't we keep "dom" here to ensure that node and dom are compatible?

Comment (SimonSchick) [reply types/node/ts3.1/tsconfig.json]: Yes and no.
I don't think they are going to be compatible in the long run and it's already causing issues (see TextEncoder/console etc.), eventually these will have to be broken up or `console` etc. need to be moved to a shared lib.

I will restore it for now but I still think it's a nasty anti pattern.

**Comment by SimonSchick** [comment]
@rbuckton if you decide to merge this, please squash :)

**Comment by SimonSchick** [comment]
@Flarna you happy? :P

File: types/node/ts3.1/tsconfig.json
Comment (Flarna) [reply types/node/ts3.1/tsconfig.json]: I' assume that a lot people don't do this on purpose; it just happens because of dependencies.

**Review by Flarna — APPROVED** [review]

**Comment by SimonSchick** [comment]
@btoueg @Flarna merged changes from https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32597 into this PR.
Also:
Cleaned up crypto file by introducing `Binary` and `BinaryLike` and `KeyLike` types to reduce type union bloat.

**Comment by Flarna** [comment]
Merge seems to be ok. I have no special opinion regarding the new crypto types.
But it seems the removal of `crypto.Credentials` results in an unhappy irc module.

Any idea why other node maintainer have not been invited for a review by @typescript-bot ?

**Comment by SimonSchick** [comment]
@Flarna according to the docs crypto.Credentials was wrong, unfortunately I can't run the node depedency tests locally, will fix soon. :)

**Comment by qm3ster** [comment]
Thank you for the `toJSON(): { type: 'Buffer', data: number[] };`
Shouldn't the `encoding` argument in `Buffer.write` be the `BufferEncoding` union declared above instead of `string`? (I mentioned this as an also in #32926, not sure if anyone saw)

**Comment by SimonSchick** [comment]
@qm3ster I guess so, would also make sense to create separate overloads for all other functions that take `encoding` as it's only required for strings.
Please create a new issue for this, not sure when I will get around to do it.

**Pull request merged**
